### ISSUE TITLE
Accessing package data while using sync()

### ISSUE THREAD

**Issue by FittinRoom** [Microsoft/cordova-plugin-code-push#262]
### Description

This is not really an issue but more a feature request if it's not already available (although, I guess it's not). I would like to be able to access the package information while using the sync method: more precisely, I need to be able to detect if an update currently analyzed/installed with sync() is mandatory or not...

In our app flow (ionic cordova), we first go to an updateCheck page, which checks for update then installs it if available. 
The problem is that, after the installation has been made (SyncStatus.UPDATE_INSTALLED) , I need to either: 1) wait for the restart if the package was marked as mandatory, 2) redirect the user to an app page if the packaged was not mandatory.

I observed that, if we just redirect the user to our app page (using $state.go(..) ) when the syncStatus is UPDATE_INSTALLED for a mandatory update, then the update is often marked as failed . I guess that the sync function is not totally finished when the page transition occurs, which leads to the destruction of the controller and the sync function never finishes. When the update is not mandatory, redirecting the user to our app page does not seems to pose any issue (package will be installed on next restart anyway)

Hence, I would like to be able to do something like:
```
var syncStatusCallback = function (syncStatus) {
        switch (syncStatus) {
          // Result (final) statuses
          case SyncStatus.UPDATE_INSTALLED:
            console.log('update installed')
            if (!package.isMandatory){
                $state.go('page');
            }
            break;
};
window.codePush.sync(syncStatusCallback, null, downloadProgress);
```

Note the 
 **if (!package.isMandatory){
                $state.go('page');
            }**
additional line.

Thanks

### Reproduction

N/A

### Additional Information

* cordova-plugin-code-push version: 
cordova-plugin-code-push 1.9.4-beta "CodePush"
code-push 2.0.1-beta "CodePushAcquisition"

* List of installed plugins:
code-push 2.0.1-beta "CodePushAcquisition"
com.phonegap.plugins.PushPlugin `3.3.0` "PushPlugin"
com.telerik.plugins.nativepagetransitions 0.6.5 "Native Page Transitions"
cordova-plugin-3dtouch 1.3.5 "3D Touch"
cordova-plugin-add-swift-support 1.6.2 "AddSwiftSupport"
cordova-plugin-appavailability 0.4.2 "AppAvailability"
cordova-plugin-badge 0.8.1 "Badges"
cordova-plugin-camera 2.4.0 "Camera"
cordova-plugin-camera-preview 0.9.0 "cordova-plugin-camera-preview"
cordova-plugin-code-push 1.9.4-beta "CodePush"
cordova-plugin-compat 1.1.0 "Compat"
cordova-plugin-console 1.0.6 "Console"
cordova-plugin-contacts 2.3.1 "Contacts"
cordova-plugin-device 1.1.6 "Device"
cordova-plugin-dialogs 1.3.3 "Notification"
cordova-plugin-file 4.3.3 "File"
cordova-plugin-file-transfer 1.6.3 "File Transfer"
cordova-plugin-geolocation 2.4.3 "Geolocation"
cordova-plugin-globalization 1.0.6 "Globalization"
cordova-plugin-inappbrowser 1.7.1 "InAppBrowser"
cordova-plugin-network-information 1.3.2 "Network Information"
cordova-plugin-photo-library 2.0.4 "Photo Library"
cordova-plugin-splashscreen 4.0.3 "Splashscreen"
cordova-plugin-statusbar 2.2.3 "StatusBar"
cordova-plugin-vibration 2.1.5 "Vibration"
cordova-plugin-whitelist 1.3.2 "Whitelist"
cordova-plugin-x-socialsharing 5.1.8 "SocialSharing"
cordova-plugin-zip 3.1.0 "cordova-plugin-zip"
cordova.plugins.diagnostic 3.6.5 "Diagnostic"
es6-promise-plugin 4.1.0 "Promise"
ionic-plugin-keyboard 2.2.1 "Keyboard"
phonegap-plugin-barcodescanner 6.0.6 "BarcodeScanner"

* Cordova version:
6.5.0

* iOS/Android/Windows version:
iOS/Android

* Does this reproduce on a debug build or release build?
both

* Does this reproduce on a simulator, or only on a physical device?
both

**Comment by max-mironov** [comment]
Hi @FittinRoom, that's a good question and here are some thoughts:
Why don't you give a try to use manual checking and installing updates? This way you will have full control over what is going on here (in fact sync is just a convenient wrapper to be used to avoid difficulties with manual installation steps).
So seems you could do the following:
1) [Asks the CodePush service whether the configured app deployment has an update available](https://github.com/Microsoft/cordova-plugin-code-push#codepushcheckforupdate).
2) [Download local package](https://github.com/Microsoft/cordova-plugin-code-push#methods-1) if exists
3) [Examine if downloaded package isMandatory or not](https://github.com/Microsoft/cordova-plugin-code-push#properties)
4) [Install local package](https://github.com/Microsoft/cordova-plugin-code-push#methods)
5) Don't forget to [call NotifyAppReady ](https://github.com/Microsoft/cordova-plugin-code-push#codepushnotifyapplicationready)in order to be protected against bad updates (this is important, otherwise your update will be rolled back)

So in short your code will look somehow like this:
```
var onError = function (error) {
    console.log("An error occurred. " + error);
};

var onInstallSuccess = function () {
    console.log("Installation succeeded.");
};

var onPackageDownloaded = function (localPackage) {
    // Install regular updates after someone navigates away from the app for more than 2 minutes
    // Install mandatory updates after someone restarts the app
    // Or just check if localPackage isManadatory or not and do your custom logic here
    localPackage.install(onInstallSuccess, onError, { installMode: InstallMode.ON_NEXT_RESUME, minimumBackgroundDuration: 120, mandatoryInstallMode: InstallMode.ON_NEXT_RESTART });
};

var onUpdateCheck = function (remotePackage) {
    if (!remotePackage) {
        console.log("The application is up to date.");
    } else {
        // The hash of each previously reverted package is stored for later use. 
        // This way, we avoid going into an infinite bad update/revert loop.
        if (!remotePackage.failedInstall) {
            console.log("A CodePush update is available. Package hash: " + remotePackage.packageHash);
            remotePackage.download(onPackageDownloaded, onError);
        } else {
            console.log("The available update was attempted before and failed.");
        }
    }
};

window.codePush.checkForUpdate(onUpdateCheck, onError);
```

and again don't forget to call NotifyAppReady in the updated version of the app:
```
var app = {
    onDeviceReady: function () {
        // Calling this function is required during the first application run after an update.
        // If not called, the application will be reverted to the previous version.
        window.codePush.notifyApplicationReady();
        // ...
    }
}
```

Please share your thoughts what do you think about it and if it could work for you.

**Comment by FittinRoom** [comment]
Hello @max-mironov ,

I totally agree with your proposed solution. To be honest, that's what I was going to set up in case the package information was not available with the sync() function.
However, it still seems strange that the sync() function can't access package information since it's displayed in the console before starting the download.. I guess it's just not exposed and that would be an interesting update in my point of view.

Thanks for your proposed solution.

**Comment by max-mironov** [comment]
Hey @FittinRoom, thank you for feedback.
Agreed that it would be good enough to have access to package object while doing sync, but frankly I believe that this is the first request we received about this. So changing the contract of `syncCallback` would probably lead to some kind of pain to those who is already using their logic to expect a single parameter from `syncCallback`, not e.g. an object containing two properties (syncStatus and package). 

For now I suppose we could postpone this and we can go ahead and implement this improvement if we have more user requests about it.

**Comment by FittinRoom** [comment]
Sure, I agree with that. However, wouldn't it be easier to add a function getInstalledPackage() (similar to getRemotePackage() and getLocalPackage() ) that would return the last package object downloaded/installed?
Not sure if that's possible in the code logic though.
Anyway, thanks for your feedback, I'm currently changing my code logic with respect to your proposed solution.
Cheers

**Comment by max-mironov** [comment]
Closing the item for now, could be reopened if get more votes for this feature request.
